packages/beacon-node/src/sync/unknownBlock.ts

```diff
@@ -825,4 +753,231 @@ export class UnknownBlockSync {
 
     return badPendingBlocks;
   }
+
+  private getMaxDownloadAttempts(): number {
+    if (this.config.getForkSeq(this.chain.clock.currentSlot) < ForkSeq.fulu) {
+      return MAX_ATTEMPTS_PER_BLOCK;
+    }
+
+    // TODO: I consider max 20 downloads per block for a supernode is enough for devnets
+    // review this computation for public testnets or mainnet
+    return Math.min(
+      20,
+      (MAX_ATTEMPTS_PER_BLOCK * this.network.custodyConfig.sampleGroups.length) / this.config.SAMPLES_PER_SLOT
+    );
+  }
+}
+
+/**
+ * Class to track active byRoots requests and balance them across eligible peers.
+ */
+export class UnknownBlockPeerBalancer {
+  private readonly peersMeta: Map<PeerIdStr, PeerSyncMeta>;
+  private readonly activeRequests: Map<PeerIdStr, number>;
+  private readonly custodyConfig: CustodyConfig;
+
+  constructor(custodyConfig: CustodyConfig) {
+    this.peersMeta = new Map();
+    this.activeRequests = new Map();
+    this.custodyConfig = custodyConfig;
+  }
+
+  onPeerConnected(peerId: PeerIdStr, syncMeta: PeerSyncMeta): void {
+    if (!this.peersMeta.has(peerId)) {
+      this.peersMeta.set(peerId, syncMeta);
+    }
+
+    if (!this.activeRequests.has(peerId)) {
+      this.activeRequests.set(peerId, 0);
+    }
+  }
+
+  onPeerDisconnected(peerId: PeerIdStr): void {
+    this.peersMeta.delete(peerId);
+    this.activeRequests.delete(peerId);
+  }
+
+  /**
+   * called from fetchUnknownBlockRoot() where we only have block root and nothing else
+   * excludedPeers are the peers that we requested already so we don't want to try again
+   * pendingColumns is empty for prefulu, or the 1st time we we download a block by root
+   */
+  bestPeerForPendingColumns(pendingColumns: Set<number> | null, excludedPeers: Set<PeerIdStr>): PeerSyncMeta | null {
+    const eligiblePeers = this.filterPeers(pendingColumns, excludedPeers);
+    if (eligiblePeers.length === 0) {
+      return null;
+    }
+
+    const sortedEligiblePeers = sortBy(
+      shuffle(eligiblePeers),
+      // prefer peers with least active req
+      (peerId) => this.activeRequests.get(peerId) ?? 0
```

Comment:
Should we be checking for max of two requests per method also?

Reply:
happening in filterPeers